tsdb/blockwriter.go

```diff
@@ -0,0 +1,129 @@
+
+package tsdb
+
+import (
+	"context"
+	"io/ioutil"
+	"math"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+// BlockWriter is a block writer that allows appending and flushing series to disk.
+type BlockWriter struct {
+	logger         log.Logger
+	destinationDir string
+
+	head      *Head
+	blockSize int64 // in ms
+	chunkDir  string
+}
+
+func durationToMillis(t time.Duration) int64 {
+	return int64(t.Seconds() * 1000)
+}
+
+// NewBlockWriter create a new block writer.
+//
+// The returned writer accumulates all the series in the Head block until `Flush` is called.
+//
+// Note that the writer will not check if the target directory exists or
+// contains anything at all. It is the caller's responsibility to
+// ensure that the resulting blocks do not overlap etc.
+// Writer ensures the block flush is atomic (via rename).
+func NewBlockWriter(logger log.Logger, dir string, blockSize int64) (*BlockWriter, error) {
+	w := &BlockWriter{
+		logger:         logger,
+		destinationDir: dir,
+		blockSize:      blockSize,
+	}
+	err := w.initHead()
+	if err != nil {
+		return nil, err
+	}
+	return w, nil
+}
+
+// initHead creates and initialises a new TSDB head.
+// blockSize should be in milliseconds.
+func (w *BlockWriter) initHead() error {
+	chunkDir, err := ioutil.TempDir(os.TempDir(), "head")
+	if err != nil {
+		return errors.Wrap(err, "create temp dir")
+	}
+	w.chunkDir = chunkDir
+
+	h, err := NewHead(nil, w.logger, nil, w.blockSize, w.chunkDir, nil, DefaultStripeSize, nil)
+	if err != nil {
+		return errors.Wrap(err, "tsdb.NewHead")
+	}
+
+	w.head = h
+	return w.head.Init(math.MinInt64)
+}
+
+// Appender returns a new appender on the database.
+// Appender can't be called concurrently. However, the returned Appender can safely be used concurrently.
+func (w *BlockWriter) Appender(ctx context.Context) storage.Appender {
+	return w.head.Appender(ctx)
+}
+
+// Flush implements the Writer interface. This is where actual block writing
+// happens. After flush completes, no writes can be done.
+func (w *BlockWriter) Flush(ctx context.Context) (ulid.ULID, error) {
+	seriesCount := w.head.NumSeries()
+	if w.head.NumSeries() == 0 {
+		return ulid.ULID{}, errors.New("no series appended, aborting")
+	}
+
+	mint := w.head.MinTime()
+	// Add +1 millisecond to block maxt because block intervals are half-open: [b.MinTime, b.MaxTime).
+	// Because of this block intervals are always +1 than the total samples it includes.
+	maxt := w.head.MaxTime() + 1
+	level.Info(w.logger).Log("msg", "flushing", "series_count", seriesCount, "mint", timestamp.Time(mint), "maxt", timestamp.Time(maxt))
+
+	compactor, err := NewLeveledCompactor(ctx,
+		nil,
+		w.logger,
+		[]int64{durationToMillis(2 * time.Hour)},
```

Comment:
Ditto, you're making assumptions here that don't belong in the TSDB code.

Reply:
Yea, I guess that is what was designed for TSDB. I guess this means, it should be part of some other package? :thinking:

Reply:
I'd say pass in the value instead.

Reply:
We could use the `w.blockSize` here instead.